### DISCUSSION TITLE
NFT 创建过程的理解与问题

### DISCUSSION THREAD

**Discussion by livelybug** [NonceGeek/Move-Camp#283] [Q&A] [answered]
一直没接触过NFT，上了<05. NFT 合约案例讲解>之后，费大劲才把NFT 创建过程基本理顺了。
以下是我的总结，为了方便表达，把[这两行代码](https://github.com/hqwangningbo/Aptos-Simple-Token/blob/main/simple_token/sources/NFTIssue.move#L119-L120)的参数换成了Alice / Bob.
刚开始是中文写的，发现英文单词较多，索性写成英文。
最后带了一个问题： [// only 1 copy for the token with property verion bigger than 0](https://github.com/aptos-labs/aptos-core/blob/9518e1be2ff471cc875385ba436132c6ead46732/aptos-move/framework/aptos-token/sources/token.move#L844) 这句注释没看懂。
请各位老师帮忙看看以下的NFT创建过程有没有错误，还有那句注释什么意思。谢谢

```rust
        init(Alice,10,1000,1000000000);
        mint(Bob);
```

*   Project owner(`Alice`) creates a token collection(`AliceCollection`) of type `struct Collections`&#x20;. Note: `AliceCollection` will be the key to a `CollectionData` under `Alice`'s resource `struct Collections`.  If Alice creates another token collection(`MonkeyCollection`), `MonkeyCollection` will be another key to a `CollectionData` in the same table as `AliceCollection`.
*   `Alice` creates a token data ID(`AliceOne`) of type `struct TokenDataId`.&#x20;
*   `AliceCollection` and `AliceOne` are linked by the collection name of type `CollectionData->name`. ~~Note:  `AliceCollection` and `AliceOne` may belong to different creators.~~ Note: The creator of `AliceCollection` or `AliceOne` may not have the signer capability to mint `AliceOne`.
*   The capability to mint `AliceOne` is assigned to `Alice` &#x20;
*   `Alice` mints one  `AliceOne`&#x20;
*   Initialize `Alice`'s `TokenStore`&#x20;
*   If `Alice`'s `TokenStore` doesn't contain `AliceOne`, insert `AliceOne` into the `TokenStore`;  Otherwise, increase `AliceOne`'s amount. &#x20;
*   Withdraw one `AliceOne` from `Alice`'s `TokenStore`
*   Initialize `Bob`'s `TokenStore`&#x20;
*   Insert one `AliceOne` into `Bob`'s `TokenStore`
*   Convert one and only one `AliceOne` to NFT:   
    + if `AliceOne`'s `property_version` == 0,  increase token `property_version` by one; 
    + if `property_version` > 0, update token properties. `// only 1 copy for the token with property verion bigger than 0`?

**Comment by qiwihui** [comment, marked as answer]
整体流程是正确的。

> if AliceOne's property_version == 0, increase token property_version by one;

这个地方也会 update token properties。

有一个注释不太确定：

> AliceCollection and AliceOne are linked by the collection name of type CollectionData->name. Note: AliceCollection and AliceOne may belong to different creators.

创建 TokenData 的时候，调用 `token::create_tokendata` 需要传入 account 的 `&signer` 并检查 account 账户下是否有 Collection，所以应该只能 Collection 的创建者调用，不太确定如何做到属于不同的创建者？  https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token/sources/token.move#L1095-L1122

**Reply by livelybug** [reply]
谢谢指正！
我一开始理解TokenDataId与Collections可能属于不同创建者的原因是：
[mutate_token_properties](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token/sources/token.move#L433) 方法的参数中有 `account`， `token_owner`，和`creator` 三个账户。其中的`account`我理解是collection创建者，`creator`是TokenDataId创建者， 所以判断TokenDataId与Collections可能属于不同创建者。
刚刚想了一下，`account`不一定必须是collection创建者，而是在创建TokenDataId时指定的有Signer Capability的账户。
